### PR TITLE
Implement hydrated components in the static build

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- This removes vite postprocessing for the static build
- Also gets hydrated components working
- Note that the normal build is unaffected.

## Testing

Fast build demo is being used as the test as of now.

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/core/build/internal.ts [hunk 1]
```diff
@@ -14,6 +14,8 @@ export interface BuildInternals {
 
 	// A mapping to entrypoints (facadeId) to assets (styles) that are added.
 	facadeIdToAssetsMap: Map<string, string[]>;
+
+	entrySpecifierToBundleMap: Map<string, string>;
```
Comment (matthewp) [inline packages/astro/src/core/build/internal.ts:18]: Add docs on what this does and is for.

File: packages/astro/src/core/build/static-build.ts [hunk 2]
```diff
@@ -28,35 +32,47 @@ export interface StaticBuildOptions {
 export async function staticBuild(opts: StaticBuildOptions) {
 	const { allPages, astroConfig } = opts;
 
+	// The pages
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:35]: Better description here

File: packages/astro/src/core/build/static-build.ts [hunk 3]
```diff
@@ -151,14 +203,34 @@ async function generatePath(path: string, opts: StaticBuildOptions, gopts: Gener
...
+		result.links = new Set<SSRElement>(
+			linkIds.map((href) => ({
+				props: {
+					rel: 'stylesheet',
+					href,
+				},
+				children: '',
+			}))
+		);
+		result.resolve = async (specifier: string) => {
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:222]: Add doc comment on what this is for / doing

File: packages/astro/src/core/build/static-build.ts [hunk 4]
```diff
@@ -167,7 +239,21 @@ async function generatePath(path: string, opts: StaticBuildOptions, gopts: Gener
 	}
 }
 
-export function vitePluginNewBuild(): VitePlugin {
+async function cleanSsrOutput(opts: StaticBuildOptions) {
+	// The SSR output is all .mjs files, the client output is not.
+	const files = await glob('**/*.mjs', {
+		cwd: opts.astroConfig.dist.pathname,
+		//ignore: ['node_modules/**'].concat(filePathsToIgnore.map((ignore) => `${ignore}/**`)),
```
Comment (matthewp) [inline packages/astro/src/core/build/static-build.ts:246]: Remove this

File: packages/astro/src/core/ssr/index.ts [hunk 5]
```diff
@@ -389,7 +286,8 @@ export async function render(renderers: Renderer[], mod: ComponentInstance, ssrO
 	// run transformIndexHtml() in dev to run Vite dev transformations
 	if (mode === 'development') {
 		const relativeURL = filePath.href.replace(astroConfig.projectRoot.href, '/');
-		html = await viteServer.transformIndexHtml(relativeURL, html, pathname);
+		console.log('TRANFORM', relativeURL, html);
```
Comment (matthewp) [inline packages/astro/src/core/ssr/index.ts:289]: whoops. I think this still can't be removed for non static builds, so this should be updated.

File: packages/astro/src/runtime/server/hydration.ts [hunk 6]
```diff
@@ -69,6 +69,9 @@ export function extractDirectives(inputProps: Record<string | number, any>): Ext
 					extracted.hydration.componentExport.value = value;
 					break;
 				}
+				case 'client:component-hydration': {
```
Comment (matthewp) [inline packages/astro/src/runtime/server/hydration.ts:72]: Add a comment here.

Comment (matthewp) [reply packages/astro/src/runtime/server/hydration.ts:72]: Also add a warning.

File: packages/astro/src/runtime/server/index.ts [hunk 7]
```diff
@@ -157,6 +157,7 @@ Did you mean to enable ${formatList(probableRendererNames.map((r) => '`' + r + '
 
 	// Call the renderers `check` hook to see if any claim this component.
 	let renderer: Renderer | undefined;
+	debugger;
```
Comment (matthewp) [inline packages/astro/src/runtime/server/index.ts:160]: Remove

File: packages/astro/src/vite-plugin-build-html/index.ts [hunk 8]
```diff
@@ -120,18 +120,14 @@ export function rollupPluginAstroBuildHTML(options: PluginOptions): VitePlugin {
...
-					const styleId = getAstroStyleId(pathname);
-					let styles = 0;
+					let styles = '';
 					for (const node of findInlineStyles(document)) {
 						if (hasAttribute(node, 'astro-style')) {
-							const style = getTextContent(node) || ' '; // If an empty node, add whitespace
-							const thisStyleId = `${styleId}/${++styles}.css`;
-							internals.astroStyleMap.set(thisStyleId, style);
-							assetImports.push(thisStyleId);
+							styles += getTextContent(node);
```
Comment (matthewp) [inline packages/astro/src/vite-plugin-build-html/index.ts:126]: this is wrong :(. Need to take whats in `main` instead.

File: packages/renderers/renderer-vue/server.js [hunk 9]
```diff
@@ -3,7 +3,7 @@ import { renderToString } from 'vue/server-renderer';
 import StaticHtml from './static-html.js';
 
 function check(Component) {
-	return !!Component['ssrRender'];
+	return !!Component['ssrRender'] || !!Component.render;
```
Comment (matthewp) [inline packages/renderers/renderer-vue/server.js:6]: This can be removed

File: packages/astro/src/core/ssr/index.ts [hunk 10]
```diff
@@ -139,6 +125,7 @@ export async function preload({ astroConfig, filePath, viteServer }: SSROptions)
 	return [renderers, mod];
 }
 
+// TODO REMOVE
```
Comment (matthewp) [inline packages/astro/src/core/ssr/index.ts:128]: Pretty sure this function can be removed and is not being used now

**Review by jonathantneal — APPROVED** [review]
Appreciate that it comes with a fast-build.

**Pull request merged**